source/slang/slang-emit-wgsl.cpp

```diff
@@ -0,0 +1,1013 @@
+    case Stage::Vertex:
+        m_writer->emit("@vertex\n");
+        break;
+
+    case Stage::Compute:
+    {
+        m_writer->emit("@compute\n");
+
+        {
+            Int sizeAlongAxis[kThreadGroupAxisCount];
+            getComputeThreadGroupSize(irFunc, sizeAlongAxis);
+
+            m_writer->emit("@workgroup_size(");
+            for (int ii = 0; ii < kThreadGroupAxisCount; ++ii)
+            {
+                if (ii != 0)
+                    m_writer->emit(", ");
+                m_writer->emit(sizeAlongAxis[ii]);
+            }
+            m_writer->emit(")\n");
+        }
+    }
+    break;
+
+    default:
+        SLANG_ABORT_COMPILATION("unsupported stage.");
+    }
+
+}
+
+// This is 'function_header' from the WGSL specification
+void WGSLSourceEmitter::emitFuncHeaderImpl(IRFunc* func)
+{
+    Slang::IRType * resultType = func->getResultType();
+    auto name = getName(func);
+
+    m_writer->emit("fn ");
+    m_writer->emit(name);
+
+    emitSimpleFuncParamsImpl(func);
+
+    // An absence of return type is expressed by skipping the optional '->' part of the
+    // header.
+    if (resultType->getOp() != kIROp_VoidType)
+    {
+        m_writer->emit(" -> ");
+        emitType(resultType);
+    }
+}
+
+void WGSLSourceEmitter::emitSimpleFuncParamImpl(IRParam* param)
+{
+    if (auto sysSemanticDecor = param->findDecoration<IRTargetSystemValueDecoration>())
+    {
+        m_writer->emit("@builtin(");
+        m_writer->emit(sysSemanticDecor->getSemantic());
+        m_writer->emit(")");
+    }
+
+    CLikeSourceEmitter::emitSimpleFuncParamImpl(param);
+}
+
+void WGSLSourceEmitter::emitTransposedMatrixType(
+    IRType *const elementType, IRMatrixType *const matrixType
+    )
+{
+    IRIntegerValue const colCountSlang {getIntVal(matrixType->getColumnCount())};
+    IRIntegerValue const rowCountSlang {getIntVal(matrixType->getRowCount())};
+    IRIntegerValue const colCountWGSL {rowCountSlang};
+    IRIntegerValue const rowCountWGSL {colCountSlang};
+    emitMatrixType(elementType, rowCountWGSL, colCountWGSL);
+}
+
+void WGSLSourceEmitter::emitMatrixType(
+    IRType *const elementType, IRIntegerValue const& rowCountWGSL,
+    IRIntegerValue const& colCountWGSL
+    )
+{
+    // WGSL uses CxR convention
+    m_writer->emit("mat");
+    m_writer->emit(colCountWGSL);
+    m_writer->emit("x");
+    m_writer->emit(rowCountWGSL);
+    m_writer->emit("<");
+    emitType(elementType);
+    m_writer->emit(">");
+}
+
+void WGSLSourceEmitter::emitStructDeclarationSeparatorImpl()
+{
+    m_writer->emit(",");
+}
+
+static bool isPowerOf2(uint32_t const n)
+{
+    return (n != 0U) && ((n - 1U) & n) == 0U;
+}
+
+void WGSLSourceEmitter::emitStructFieldAttributes(
+    IRStructType * structType, IRStructField * field
+    )
+{
+    // Tint emits errors unless we explicitly spell out the layout in some cases, so emit
+    // offset and align attribtues for all fields.
+    IRSizeAndAlignmentDecoration *const sizeAndAlignmentDecoration {
```

Comment:
no initializer list or function call style initialization if we can avoid it because it is not as readable as `T v = value;`

Reply:
Done. (It's a habit from working on safety-critical code -- I think the brace style init forces you to be more explicit about integers in some cases.)